### DISCUSSION TITLE
Usefulness of a (:if) stanza

### DISCUSSION THREAD

**Discussion by F-Loyer** [ocaml/dune#9917] [Ideas]
Many stanzas supports an `(:include)` argument, but if we simply want a conditional argument, we need to declare a pair of rules like:

```
(executable
 (name pg_check)
 (public_name pg_check)
 (link_flags (:include linkflags.sexp))
 (libraries ctypes ctypes.foreign cmdliner pg_query))

(rule
  (target linkflags.sexp)
  (enabled_if (= %{os_type} Win32))
  (action (with-outputs-to %{target}
    (echo "(-cclib -Wl,-Wl,--export-all-symbols)"))))
(rule
  (target linkflags.sexp)
  (enabled_if (<> %{os_type} Win32))
  (action (with-outputs-to %{target}
    (echo "()"))))
```

With an `(:if)` argument, we can imagine something like:

```
(executable
  (name pg_check)
  (public_name pg_check)
  (link_flags (:if (= %{os_type} Win32) (-cclib -Wl,-Wl,--export-all-symbols)))
  (libraries ctypes ctypes.foreign cmdliner pg_query))
```
Ideally, the `(:if)` could also be used in the form `(:if (condition) (if_true_value) (if_false_value))`, and the `if_true_value` or `if_false_value` could also be written as a nested `(:if)` argument (https://github.com/lukstafi/ocaml-gccjit/blob/master/lib/dune gives a more complex example where nested `:if` would be needed.

We can also go further with `if_true_value` or `if_false_value` using `(:include)` (use case: having multiple static included files, and the `:if` select the right one. But I admit it to be less usefull).

**Comment by rgrinberg** [comment]
Thanks for the suggestion. It would indeed be nice to have something like this and we've added something similar with `cond` for lock directory selection. The only problem is that I'm really not a fan of having to annotate every special value with `:if`. Unfortunately, I don't have an alternative to propose either, so I suppose this is fine. @gridbugs do you have an opinion on this?

**Comment by F-Loyer** [comment]
I may perhaps try to propose a PR… but I would have to discover the dune architecture before being able to propose something. A grep « :include » may help finding the spots where to include the « :if » implementation (once it is developed).--Frédéric LoyerLe 3 mars 2024 à 16:28, Rudi Grinberg ***@***.***> a écrit :﻿
Thanks for the suggestion. It would indeed be nice to have something like this and we've added something similar with cond for lock directory selection. The only problem is that I'm really not a fan of having to annotate every special value with :if. Unfortunately, I don't have an alternative to propose either, so I suppose this is fine. @gridbugs do you have an opinion on this?

**Reply by rgrinberg** [reply]
Sure, you're welcome to give it a hand. It's certainly something we'd like to have in one form or the other.

**Comment by gridbugs** [comment]
Sounds like a useful addition. If you haven't already, check out this page in our docs which states that we intend to add such a feature one day: https://dune.readthedocs.io/en/stable/reference/ordered-set-language.html

> Note that inside an ordered set, a list’s first element cannot be an atom except if it starts with - or :. The reason for this is that we’re planning to add simple programmatic features in the future so that one may write:
`(flags (if (>= %{ocaml_version} 4.06) ...))`

@rgrinberg I don't understand what you mean by:
> The only problem is that I'm really not a fan of having to annotate every special value with `:if`.

Can you elaborate on what you mean by "special value".

**Reply by rgrinberg** [reply]
Ah right, I forgot about that restriction.

My initial suggestion was that it wouldn't be possible to have `(if a b c)` because `if` would be treated as a literal. I see it's no longer needed because of the restriction you've linked. I actually don't remember when/where it was implemented. Do we have tests documenting this limitation?

**Reply by gridbugs** [reply]
Doesn't look like we have tests but it's implemented here: https://github.com/ocaml/dune/blob/edd1192b704208d84c31d2942808e74a6ad8eca9/src/dune_lang/ordered_set_lang.ml#L83

**Reply by rgrinberg** [reply]
I see. Could you add some tests for this?

**Reply by gridbugs** [reply]
https://github.com/ocaml/dune/pull/10226
